### PR TITLE
All the code

### PR DESCRIPTION
**Pull request by rosskarchner**
There's a few rough edges and missing features:
- it will silently overwrites existing build and release zip files (if you re-use a build label, for example)
- There's no code in place that supports a persistent 'lib' directory (for apps and modules that will be deployed on their own schedule, separate from the project release cycle)
- setup.py needs love
- the --append-wsgi option was built to support newrelic, but it turns out the newrelic code needs _pre_-pended.
- "paths.json" is thus far undocumented, but it should be pretty easy to figure out if you look at dfd_paths.py

... but I think this is a good conversation starter ;)

### CODE REVIEW

File: no_drama/__main__.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+templates_dir = os.path.join(this_directory, 'templates')
+build_lib = os.path.join(this_directory, 'build_lib')
+build_files = os.path.join(this_directory, 'build_files')
+
+
+def save_wheels(destination, packages=None, requirements_file=None,
+                find_links=None):
+    call_list = ["pip", "wheel", "--wheel-dir=%s" % destination]
+
+    if packages is not None:
```
Comment (vickumar1981) [inline no_drama/__main__.py]: use truthiness, i.e.

```
if not packages:
```

File: no_drama/__main__.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        call_list += packages
+
+    if find_links is not None:
```
Comment (vickumar1981) [inline no_drama/__main__.py]: ditto.

also, you can re-write this as an if-comprehension:

call_list += ['--find-links=%s' % find_links] if find_links else list()

File: no_drama/__main__.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        call_list += ['--find-links=%s' % find_links]
+
+    if requirements_file is not None:
```
Comment (vickumar1981) [inline no_drama/__main__.py]: same.

File: no_drama/build_lib/dfd_settings.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import os
+import dfd_paths
+
+real_settings = os.environ.get('REAL_DJANGO_SETTINGS')
+
+if real_settings is not None:
```
Comment (vickumar1981) [inline no_drama/build_lib/dfd_settings.py]: truthiness.

File: no_drama/__main__.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    subprocess.call(call_list)
+
+
+def get_file_from_dir(dir, filename, mode='r'):
+    full_path = os.path.join(dir, filename)
+    return open(full_path, mode)
+
+
+def get_template(filename, mode='r'):
+    template_file = get_file_from_dir(templates_dir, filename, mode=mode)
```
Comment (vickumar1981) [inline no_drama/__main__.py]: what happens if the file doesn't exist?  i feel like get_file_from_dir should handle that exception.

File: no_drama/__main__.py (hunk 1 above)
Comment (willbarton) [inline no_drama/__main__.py]: That will only check `packages.__nonzero__`. 

I like the more explicit `packages is not None`. It's clear exactly what the case is.

File: no_drama/__main__.py (hunk 5 above)
Comment (rosskarchner) [inline no_drama/__main__.py]: I think failing-with-an-exception probably IS the right thing-- these are files that are bundled with no-drama itself.

File: no_drama/build_lib/dfd_paths.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        paths.update(configured_paths)
+
+
+def abs_path(name):
+    relpath = paths[name]
+    if relpath is not None and bool(relpath) is True:
+        joined = os.path.join(root, relpath)
+        return os.path.normpath(joined)
+    else:
+        return None
```
Comment (vickumar1981) [inline no_drama/build_lib/dfd_paths.py]: don't need that else block.  if you don't return anything from a python function, None is the default return value, i believe, but it's a bad idea to return None's or null's from a function, or to pass them in as parameters.

File: no_drama/build_lib/activate_phase2.py [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import os
+
+from dfd_paths import root, update_symlink
+
+if update_symlink is not None:
```
Comment (vickumar1981) [inline no_drama/build_lib/activate_phase2.py]: same truthiness.

File: no_drama/__main__.py (hunk 5 above)
Comment (vickumar1981) [inline no_drama/__main__.py]: i would just wrap that line with a block:

```
fname = get_file_from_dir(...)
if os.path.isfile(fname):
   ...
else:
  raise SomeCustomException
```

either that, or you should have this in a try..except.

File: no_drama/__main__.py (hunk 1 above)
Comment (vickumar1981) [inline no_drama/__main__.py]: the line above will check for not none.  it is extraneous in python to do a check for not None.  it's simpler and cleaner the way i presented above.

Comment (willbarton) [inline no_drama/__main__.py]: It will check for `not None`, as well as other possible cases that fail a boolean where we'd either want to handle it differently or know why `packages` was a completely unexpected type. The potential for unintended side-effects suggests to me that leaving out the explicit `is [not] None` in all of these cases is not simpler or cleaner.

[PEP8 also warns against test `if foo` as opposed to `if foo is not None`.](https://www.python.org/dev/peps/pep-0008/#programming-recommendations)

File: no_drama/__main__.py (hunk 5 above)
Comment (willbarton) [inline no_drama/__main__.py]: I think it'd make more sense to `try`/`except` wherever the error should be appropriately handled.

File: no_drama/__main__.py (hunk 1 above)
Comment (vickumar1981) [inline no_drama/__main__.py]: in this case, i think we expect packages to be a list.  so you don't really have to worry about the boolean value of False here.

In fact, if you do pass in a boolean, the above code will still fail.

You can't use the += operator with a list and a boolean.

That operator tells me packages HAS to be a list.

Comment (willbarton) [inline no_drama/__main__.py]: In this case, I'd suggest `packages` default to `[]`, and then there's no need for the check (and if it's not a list, an appropriate exception will be raised).

But generally, I'd recommend strongly against `if foo` in place of `if foo is not None`.

File: no_drama/build_lib/dfd_paths.py (hunk 6 above)
Comment (rosskarchner) [inline no_drama/build_lib/dfd_paths.py]: I wanted to be explicit about returning None here-- someone eyeballing the function should be able to know that None is an expected return-- I don't think it's obvious, without the else.

I'd like to know more about returning None being bad form, though. In this case, the obvious alternative would be empty strings, which seems... sloppier, to me.

Comment (vickumar1981) [inline no_drama/build_lib/dfd_paths.py]: I think Robert Martin talks about this with nulls in java.  The problem is it is unclear what none signals.  Is it an error?  Does it just mean that it can be safely ignored?  Different programmers use none and null in different contexts to mean different things causing ambiguity.

This is one reason you'll see the error none type has no attribute...or a null pointer exception in java.

File: no_drama/__main__.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+
+    build_parser.add_argument('name', help="name of this project")
+
+    build_parser.add_argument('label', help="a label for this build-- "
+                        "maybe a build ID or version number")
+
+    build_parser.set_defaults(func=stage_bundle)
+
+    # release parser arguments
+    release_parser.add_argument('build_zip', help='path to a zip file'
```
Comment (vickumar1981) [inline no_drama/__main__.py:182]: You could create a dictionary and iterate it and call add arguments in a loop, reducing the lines of code.

File: no_drama/__main__.py (hunk 3 above)
Comment (rosskarchner) [inline no_drama/__main__.py]: 'Writing Idiomatic Python' recommends strict None comparisons when checking function arguments that default to None. (but trusting truthiness otherwise)

In most of these cases, the difference seems rather academic, though. Seems like maybe we need to start work on a style guide ;)

File: no_drama/build_lib/dfd_settings.py (hunk 4 above)
Comment (rosskarchner) [inline no_drama/build_lib/dfd_settings.py]: agreed, I'll make that change

File: no_drama/__main__.py (hunk 3 above)
Comment (willbarton) [inline no_drama/__main__.py]: Open an issue in https://github.com/cfpb/back-end! :grin:

[Google's Python Style Guide](https://google.github.io/styleguide/pyguide.html#True/False_evaluations) also mentions this as an exception to trusting implicit falseness, for reasons of ambiguity.

Comment (vickumar1981) [inline no_drama/__main__.py]: Right.  That's 100% correct.

But, the function arguments don't NEED to default to None.

```
def save_wheels(destination, packages=None,
    requirements_file=None,
    find_links=None):
   ...
```

That's the function as is.  But, let's say I rewrote that to simply:

```
def save_wheels(destination, packages=list(),
    requirements_file='',
    find_links=False):
```

Now, I'm being clear about what arguments I expect to my function, and I don't have to use the is None is Not None comparisons according to both PEP8 and 'Writing Idiomatic Python', I can use truthiness b/c I didn't rely on passing in a None value into my function.  It makes the code easier to comprehend for person xyz 10 years from now.

File: no_drama/__main__.py (hunk 5 above)
Comment (rosskarchner) [inline no_drama/__main__.py]: I'm factoring these two functions out-- an earlier version used templates everywhere, and now there's only one. I think I'll be able to convert that to a plain old file soon, too.

File: no_drama/__main__.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+def stage_bundle(cli_args):
+    staging_dir = tempfile.mkdtemp()
+    build_dir = os.path.join(staging_dir, cli_args.label)
+
+    # these are wheels needed during activation
+    bootstrap_wheels = ['virtualenv','pip','wheel','distribute']
+    bootstrap_wheels_destination = os.path.join(build_dir,'bootstrap_wheels')
+    save_wheels(packages=bootstrap_wheels, destination=bootstrap_wheels_destination)
+
+    # move just the wheels we want into the bundle dir
```
Comment (vickumar1981) [inline no_drama/__main__.py]: i think this method can be broken down a bit into smaller sub-routines.

each commented block really should just be it's own function:

```
    def activate_wheels(...)
    def move_wheels_to_bundle_dir(...)
    def copy_django_proj_to_bundle_dir(...)
    def make_lib_directory(...)
    def generate_activate_shell_file(...)
```

then you can just get rid of the comments themselves.

File: no_drama/__main__.py (hunk 2 above)
Comment (rosskarchner) [inline no_drama/__main__.py]: the "find links" stuff is not longer used, so I've cut it out.

File: no_drama/__main__.py (hunk 9 above)
Comment (rosskarchner) [inline no_drama/__main__.py]: That's a good call

File: no_drama/__main__.py (hunk 1 above)
Comment (vickumar1981) [inline no_drama/__main__.py]: https://www.python.org/dev/peps/pep-0008/#programming-recommendations

File: no_drama/__main__.py (hunk 3 above)
Comment (vickumar1981) [inline no_drama/__main__.py]: There is an exception to this that I think it is important to understand.

If packages or requirements were mutable inside the save_wheels function, then it makes sense to default them to None.  But, in this case they are not changed within that function scope, which is why defaulting them to None doesn't make sense.

**Comment by willbarton** [comment]
Should this get merged since we appear to actually be using this? :)

**Comment by rosskarchner** [comment]
I'm going to merge now-- we should make sure deployments use the 'cfpb' version from now on
